numpy/core/src/multiarray/mapping.c

```diff
@@ -285,6 +301,11 @@ prepare_index(PyArrayObject *self, PyObject *index,
     int i;
     npy_intp n;
 
+    PyObject *index = maybe_apply_callable(self, arg);
```

Comment:
Comment a bit from the sidelines, but this will slow down every `getitem` - for a rare feature such as this, I'd suggest to put it in the failure path after `unpack_indices` (or perhaps work inside `unpack_indices`)

Reply:
```python
In [1]: import numpy as np; arr = np.arange(4)

In [2]: %timeit arr[0]
```

(main): 47.7 ns ± 0.411 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
(PR): 47.6 ns ± 0.0899 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

Fair concern, but it doesn't seem to make a difference.

Reply:
I suspect that already hits a fast path that avoids this code entirely.

Reply:
OK, thanks for checking!

Reply:
I would be surprised if a callable check is not super fast.  We could move this into `unpack_indices` which would allow pushing it to after some of the fast-paths.  OTOH, we may need to guard against recursion then.

I am not quite certain if there is an exact-integer fast-path earlier (older Python versions had this in any case, but newers maybe not.  Testing `arr[0,]` or `arr[:]` might make sense)